src/hotspot/share/runtime/interfaceSupport.inline.hpp

```diff
@@ -201,35 +194,24 @@ class ThreadInVMfromNative : public ThreadStateTransition {
   ResetNoHandleMark __rnhm;
  public:
   ThreadInVMfromNative(JavaThread* thread) : ThreadStateTransition(thread) {
-    trans_from_native(_thread_in_vm);
+    transition_from_native(thread, _thread_in_vm);
   }
   ~ThreadInVMfromNative() {
-    assert(_thread->thread_state() == _thread_in_vm, "coming from wrong thread state");
     // We cannot assert !_thread->owns_locks() since we have valid cases where
     // we call known native code using this wrapper holding locks.
-    _thread->check_possible_safepoint();
-    // Once we are in native vm expects stack to be walkable
-    _thread->frame_anchor()->make_walkable(_thread);
-    OrderAccess::storestore(); // Keep thread_state change and make_walkable() separate.
-    _thread->set_thread_state(_thread_in_native);
+    transition_from_vm(_thread, _thread_in_native);
   }
 };
 
 
 class ThreadToNativeFromVM : public ThreadStateTransition {
  public:
   ThreadToNativeFromVM(JavaThread *thread) : ThreadStateTransition(thread) {
-    // We are leaving the VM at this point and going directly to native code.
-    // Block, if we are in the middle of a safepoint synchronization.
```

Comment:
Is this comment no longer correct?

Reply:
So the sentence "We are leaving the VM at this point and going directly to native code." is correct but I'm not sure what's the point of adding it. Otherwise we would have to add it for all transitions. As for "Block, if we are in the middle of a safepoint synchronization." it was true before when we used transition(), but now we just go directly to native since the safepoint will be able to progress without the need to block (equivalent to what we do in ~ThreadInVMfromNative).

Reply:
I hadn't realized we had such asymmetry with the two different VM->native transitions. Though the circumstances of the two are somewhat different.

But I can't conjure up any bad scenarios here given any safepoint/handshake is racing with this transition anyway.